spices/SPICE-0010-overhauled-mapping-listing-typechecks.adoc

```diff
@@ -0,0 +1,432 @@
+
+=== Surrogate objects
+
+Underneath the hood, casting a listing or mapping returns a _new_ listing/mapping.
+
+This new object is a surrogate of the original object.
+It behaves just like the original, except its members receive type casts too.
+This type cast may or may not throw an error.
+
+Given this:
+
+[source,pkl]
+----
+listing = new Listing { "hello" }
+
+listingOfNumbers: Listing<Int> = listing
+----
+
+`"hello"` is not an `Int`, so this program should fail.
+Because typechecking is deferred, this failure should not occur until `listing[0]` is evaluated.
+
+Effectively, `listing` and `listingOfNumbers` are _different_ in-memory objects; one which has type assertions at the end of each member, and one which does not.
+This difference is opaque to users; these two objects have the same hash code, and same equality semantics.
+
+In the amends chain, `listingOfNumbers` has the same parent as `listing`. It delegates its member lookups to `listing`, and adds a typecheck at the end of each lookup.
+
+In the following scenario:
+
+[source,pkl]
+----
+listing0 = new Listing { 1; 2; 3 }
+
+listing1 = (listing0) {
+  super[0]
+}
+
+listing2: Listing<Int> = listing1
+----
+
+`listing2` is a surrogate of `listing1`, so its parent is still `listing0`.
+Thus, `super[0]` is a lookup of the first member of `listing0`.
+
+=== Operators: Type cast (`as`) vs. Type test (`is`)
+
+The `as` and `is` operators behave differently.
+
+The `is` operator is a type test, and continues to be an _eager_ check.
+
+This snippet deeply evaluates each member, and asserts its result type.
+
+[source,pkl]
+----
+local people = new Listing {
+  new Person { name = "Sandra" }
+  new Person { name = "Zoe" }
+}
+
+isPeople = people is Listing<Person> // <1>
+----
+<1> Each member of `people` is evaluated, and asserted to be a `Person`.
+
+On the other hand, the `as` operator is a type cast, and it behaves just like a typecheck does as described in xref:_surrogate_objects[surrogate objects].
+
+[source,pkl]
+----
+local people = new Listing {
+  new Person { name = "Sandra" }
+  new Person { name = "Zoe" }
+}
+
+myPeople = people as Listing<Person> // <1>
+----
+<1> `myPeople` is a different object an is a surrogate of `people`.
+
+=== `new` with explicit parent
+
+The `new` keyword can be given a type, e.g. `new Listing<Person> {}`.
+
+Currently, this type has no influence on typechecking (see https://github.com/apple/pkl/issues/405[#405]).
+
+The following program executes without a thrown error:
+
+[source,pkl]
+----
+people = new Listing<Person> { 1; 2; 3 }
+----
+
+As part of this change, the above snippet is similarly lazily checked; the member gets casted to `Person` when evaluated.
+
+==== Amending `Listing` and `Mapping`
+
+Amended members of a `Listing` or `Mapping` do not receive type casts.
+
+[source,pkl]
+----
+listing: Listing<Int> = new { 1; 2; 3 }
+
+listing2 = (listing) { "hello" } // <1>
+----
+<1> Okay; result is `new Listing { 1; 2; 3; "hello" }`
```

Comment:
Conceptually, this makes sense to me:
We append a string to a Listing<Int> and get a listing containing ints and strings.

However, the following makes less sense to me:
```
local listing1: Listing<Int> = new { 1 }
local listing2: Listing<String> = listing1
local listing3 = (listing2) { "2" } // append a string to a Listing<String>
first = listing3[0] // we got an int (and no error)
```

Reply:
Hm, there must be a bug, then. I would expect this to error. Good catch!

Reply:
Fix: https://github.com/apple/pkl/pull/725